File: Java-Core/src/main/java/prompto/value/IntegerRange.java
package prompto.value;

import prompto.intrinsic.PromptoRange;
import prompto.type.IntegerType;


public class IntegerRange extends RangeBase<Integer> {

	static class PromptoIntegerRange extends PromptoRange<Integer> {

		public PromptoIntegerRange(Integer low, Integer high) {
			super(low, high);
		}
		
		@Override
		public Integer getItem(long item) {
			java.lang.Long result = low.longValue() + item - 1;
			if(result>high.longValue())
				throw new IndexOutOfBoundsException();
			return new Integer(result);
		}
		
		@Override
		public long length() {
			return 1L + high.longValue() - low.longValue();
		}
			
		@Override
		public PromptoIntegerRange slice(long first, long last) {
			last = adjustLastSliceIndex(last);
			return new PromptoIntegerRange(getItem(first), getItem(last));
		}


	}
	
	public IntegerRange(Integer left, Integer right) {
		this(new PromptoIntegerRange(left, right));
	}

	public IntegerRange(PromptoRange<Integer> range) {
		super(IntegerType.instance(), range);
	}

	@Override
	public long length() {
		return 1 + getHigh().longValue() - getLow().longValue();
	}

	@Override
	public int compare(Integer o1, Integer o2) {
		return o1.compareTo(o2);
	}
	
	@Override
	public RangeBase<Integer> newInstance(PromptoRange<Integer> range) {
		return new IntegerRange(range);
	}
	

}

File: Java-CodeServer/src/main/resources/js/prompto/value/IntegerRange.js
var Range = require("./Range").Range;
var Integer = require("./Integer").Integer;
var IntegerType = null;

exports.resolve =function() {
    IntegerType = require("../type/IntegerType").IntegerType;
};

function IntegerRange(left, right) {
	Range.call(this, IntegerType.instance, left, right);
	return this;
}

IntegerRange.prototype = Object.create(Range.prototype);
IntegerRange.prototype.constructor = IntegerRange;

IntegerRange.prototype.size = function() {
	return 1 + this.high.IntegerValue() - this.low.IntegerValue();
};


IntegerRange.prototype.getItem = function(index) {
	var result = this.low.IntegerValue() + index - 1;
	if(result>this.high.IntegerValue()) {
		throw new IndexOutOfBoundsException();
	}
	return new Integer(result);
};

IntegerRange.prototype.newInstance = function(left, right) {
	return new IntegerRange(left, right);
};


exports.IntegerRange = IntegerRange;